### PR TITLE
merge to develop

### PR DESCRIPTION
**Pull request by Abdusaid10**

### CODE REVIEW

**Comment by maelfosso** [comment]
# In review by Mael FOSSO

File: spec/grid_spec.rb [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+      expect(result).not_to be_nil
+    end
+  end
+  describe '#put_piece' do
+    it 'works' do
+      grid = Grid.new
+      dbl=double("Put_piece", :index=>3, :symb =>'x')
+      result=grid.put_piece(dbl.index, dbl.symb)
+      expect(result).not_to be_nil
+    end
```
Comment (maelfosso) [inline spec/grid_spec.rb]: Nothing here is showing that the piece is into the Grid even if is not returning nil
What I purpose is to check whether or not the board at that position, contains the value :X

**Comment by maelfosso** [comment]
Hello coders,

You really write a nice code. But before moving forward, there are some suggestions I want you to take in consideration.

- [ ] Please, I suggest you move the **spec** folder into the folder **tic-tac-toe** at the same level than **lib** and **bin**

- [ ] You should use symbol `:X / :O` instead of character 'x' or 'o'.

- [ ] You're testing the position **1 - 5 - 9** four times. Why ? Same thing for **1 - 2 - 3**.

Could you make those changes and submit for a new code review request?
Don't hesitate to write to me on Slack if you've any questions.

**Comment by bellom** [comment]
## Currently reviewing this PR

**Comment by Abdusaid10** [comment]
> Hello coders,
> 
> You really write a nice code. But before moving forward, there are some suggestions I want you to take in consideration.
> 
> * [ ]  Please, I suggest you move the **spec** folder into the folder **tic-tac-toe** at the same level than **lib** and **bin**
> * [ ]  You should use symbol `:X / :O` instead of character 'x' or 'o'.
> * [ ]  You're testing the position **1 - 5 - 9** four times. Why ? Same thing for **1 - 2 - 3**.
> 
> Could you make those changes and submit for a new code review request?
> Don't hesitate to write to me on Slack if you've any questions.

Thank you for your review. But why do we need to change character 'x' or 'o' to symbol ':X / :O'?.

**Comment by mcrd25** [comment]
# Currently under review by TSE [Maya Douglas](@mcrd25)

**Comment by mcrd25** [comment]
> > Hello coders,
> > You really write a nice code. But before moving forward, there are some suggestions I want you to take in consideration.
> > 
> > * [ ]   Please, I suggest you move the **spec** folder into the folder **tic-tac-toe** at the same level than **lib** and **bin**
> > * [ ]   You should use symbol `:X / :O` instead of character 'x' or 'o'.
> > * [ ]   You're testing the position **1 - 5 - 9** four times. Why ? Same thing for **1 - 2 - 3**.
> > 
> > Could you make those changes and submit for a new code review request?
> > Don't hesitate to write to me on Slack if you've any questions.
> 
> Thank you for your review. But why do we need to change character 'x' or 'o' to symbol ':X / :O'?.

Hi @Abdusaid10, the x and o are never going to change so it is best that they are represented as symbols; strings are mutable in Ruby. Does this make sense now? 😄

**Comment by mcrd25** [comment]
Hi @Abdusaid10 and @emmanuelkamala, you seem to have conflicts.

Is this the correct pull request or is this outdated?

File: tic-tac-toe/spec/grid_spec.rb [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+      result2=grid.put_piece(dbl2.index.to_i - 1, dbl2.symb)
+
+      dbl3=double("Put_piece", :index=>7, :symb =>'x')
+      result3=grid.put_piece(dbl3.index.to_i - 1, dbl3.symb)
+
+      result = grid.finished?
+      expect(result).to be true
+    end
+
+    it 'returns true if all marks on the secondary diagonal are (O)s' do
```
Comment (mcrd25) [inline tic-tac-toe/spec/grid_spec.rb:126]: What about when you expect it to return `false`?

It is best practice to cover both negative and positive scenarios for each method. Otherwise, how does the person running test know if you method actually works the way you expect it to?

File: tic-tac-toe/spec/grid_spec.rb [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+      result2=grid.put_piece(dbl2.index.to_i - 1, dbl2.symb)
+
+      dbl3=double("Put_piece", :index=>7, :symb =>'o')
+      result3=grid.put_piece(dbl3.index.to_i - 1, dbl3.symb)
+
+      result = grid.finished?
+      expect(result).to be true
+    end
+
+    it 'returns true if all the top marks are (X)s' do
```
Comment (mcrd25) [inline tic-tac-toe/spec/grid_spec.rb:158]: What about when you expect it to return `false`?

It is best practice to cover both negative and positive scenarios for each method. Otherwise, how does the person running test know if you method actually works the way you expect it to?

File: tic-tac-toe/spec/grid_spec.rb [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+      dbl2=double("Put_piece", :index=>2, :symb =>'x')
+      result2=grid.put_piece(dbl2.index.to_i - 1, dbl2.symb)
+
+      dbl3=double("Put_piece", :index=>3, :symb =>'x')
+      result3=grid.put_piece(dbl3.index.to_i - 1, dbl3.symb)
+
+      result = grid.finished?
+      expect(result).to be true
+    end
+    it 'returns true if all the top marks are (O)s' do
```
Comment (mcrd25) [inline tic-tac-toe/spec/grid_spec.rb:189]: What about when you expect it to return `false`?

It is best practice to cover both negative and positive scenarios for each method. Otherwise, how does the person running test know if you method actually works the way you expect it to?

File: tic-tac-toe/spec/grid_spec.rb [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+      dbl2=double("Put_piece", :index=>2, :symb =>'o')
+      result2=grid.put_piece(dbl2.index.to_i - 1, dbl2.symb)
+
+      dbl3=double("Put_piece", :index=>3, :symb =>'o')
+      result3=grid.put_piece(dbl3.index.to_i - 1, dbl3.symb)
+
+      result = grid.finished?
+      expect(result).to be true
+    end
+    it 'returns true if all the middle marks are (X)s' do
```
Comment (mcrd25) [inline tic-tac-toe/spec/grid_spec.rb:221]: At this point It think you get it 😛

**Review by mcrd25 — CHANGES_REQUESTED** [review]
# Good Job! Almost there! 👍 

Hi @Abdusaid10 and @emmanuelkamala,

This is good so far and meets most requirements. However, I'd liketo address following
- As you can see with my review comments you mostly cover positive scenarios in tests you should cover both positive and negative. So if you expect it to be `true` one time you should have a test that expects a method to be `false`
- This is a collaborative project and I noticed commits are from @Abdusaid10. But I guess this is simply for this branch and that other branches more accurately affect the collaborative nature of this project, so this is fine! 😅 
- Good job implementing **GitFlow**. 😄 

*If you disagree with my requests please reach out and let's have a dialogue about it 😄*

***Please makes changes and submit another review request.
If you have any questions, let me know here or on slack and I will be glad to clarify with you. You can also reach me and connect with me on my social media.***

*Happy coding 😄,*
**[--Maya](https://github.com/mcrd25)**
[GitHub](https://github.com/mcrd25) [Twitter](https://twitter.com/mcrd25) [linkedin](https://www.linkedin.com/in/mayadouglas/)

**Comment by Abdusaid10** [comment]
> > > Hello coders,
> > > You really write a nice code. But before moving forward, there are some suggestions I want you to take in consideration.
> > > 
> > > * [ ]    Please, I suggest you move the **spec** folder into the folder **tic-tac-toe** at the same level than **lib** and **bin**
> > > * [ ]    You should use symbol `:X / :O` instead of character 'x' or 'o'.
> > > * [ ]    You're testing the position **1 - 5 - 9** four times. Why ? Same thing for **1 - 2 - 3**.
> > > 
> > > Could you make those changes and submit for a new code review request?
> > > Don't hesitate to write to me on Slack if you've any questions.
> > 
> > 
> > Thank you for your review. But why do we need to change character 'x' or 'o' to symbol ':X / :O'?.
> 
> Hi @Abdusaid10, the x and o are never going to change so it is best that they are represented as symbols; strings are mutable in Ruby. Does this make sense now?

Yeah, now it is clear. Thank you for the review.

**Comment by bellom** [comment]
## Currently reviewing this PR

**Review by bellom — APPROVED** [review]
Hi @Abdusaid10 @mcrd25,
_Great work making all the necessary changes from the previous review, i check if specs are failing if i break your code in the methods, yes it did fail, awesome.
And also indentation and spacing are all good_
## Congratulation Guys :+1: 
_Go ahead and merge to master and don't forget to submit project completion form.
Goodluck in your next project :slightly_smiling_face:_

**Comment by emmanuelkamala** [comment]
Thank you!
